File: problems/1071-greatest-common-divisor-of-strings/solutions/solution.php
function gcdOfStrings(string $str1, string $str2): string {
    // If concatenations in both orders differ, no common divisor exists
    if ($str1 . $str2 !== $str2 . $str1) return "";

    // The GCD string length mirrors numeric GCD of the two lengths
    return substr($str1, 0, gcd(strlen($str1), strlen($str2)));
}

function gcd(int $a, int $b): int {
    while ($b !== 0) {
        [$a, $b] = [$b, $a % $b];
    }
    return $a;
}

File: problems/1071-greatest-common-divisor-of-strings/solutions/solution.rb
def gcd_of_strings(str1, str2)
  # If concatenations in both orders differ, no common divisor exists
  return "" if str1 + str2 != str2 + str1

  # The GCD string length mirrors numeric GCD of the two lengths
  str1[0, gcd(str1.length, str2.length)]
end

def gcd(a, b)
  a, b = b, a % b while b != 0
  a
end
